ion-c-sys/src/result.rs

```diff
@@ -10,11 +10,66 @@ use std::fmt;
 use std::num::TryFromIntError;
 
 /// IonC Error code and its associated error message.
+///
+/// `position` is populated when errors come from readers. See [`Position`] for
+/// more information.
 #[derive(Copy, Clone, Debug, PartialEq)]
 pub struct IonCError {
     pub code: i32,
     pub message: &'static str,
     pub additional: &'static str,
+    pub position: Option<Position>,
+}
+
+/// Represents a position in a data source. For example, consider a file
+/// containing Ion data that is being parsed using an [`IonCReader`].
+///
+/// `bytes` will always be hydrated while `lines` and `offset` will only be
+/// populated for text readers.
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub struct Position {
+    pub bytes: i64,
+    pub line: Option<i32>,
+    pub offset: Option<i32>,
+}
```

Comment:
I am probably going to sound nitpicking here but when I see all these options here, that really makes me wonder if `Position` is really an `enum`.  Maybe something like:

```rust
/// Represents a position in a stream/file by line and column.
pub struct LineAndColumn(usize, usize);

/// Possible position information from Ion C.
pub enum Position {
  Unknown,
  /// Number of bytes from the start of the stream/file
  Offset(usize),
  OffsetLineColumn(usize, LineAndColumn),
}
```

FWIW, I went through this evolution of problem on partiql/partiql-lang-rust#8

This has the benefit of collapsing all your `Option` modalities into one thing.  Though this factoring may make expressing logic around this worse, so feel free to push back.

Reply:
I'll make this change.